tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,296 @@ static bool HasCanonicalDimensionNumbers(
+    batch_rank_vec.push_back(out_batch_rank);
+  }
+  if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                  batch_rank_vec.begin()) &&
+      batch_rank_vec.size() > 1)
+    return false;
+  return true;
+}
+// Convert mhlo.einsum op into linalg.generic
+// Algorithm in general 3 steps:
+
+// Step1) Dissect entire einsum_config to different operands
+// e.g f("abc,cd->abd") = {lhs:["abc"], rhs:["cd"], out:["abd"]}
+
+// Step2) Split up the string into vector of the elements
+// e.g {lhs:["abc"], rhs:["cd"], out:["abd"]} = {lhs:["a","b","c"],
+// rhs:["c","d"], out:["a","b","d"]}
+
+// Step3) Convert the vector into data access
+// patern represented by affineMaps with affineDimensions e.g
+// {lhs:["a","b","c"], rhs:["c","d"], out:["a","b","d"]} = {lhs:[d0,d1,d2],
+// rhs:[d2,d3], out:[d0,d1,d3]}
+
+// Note: Step2 is especially important to for batch
+// dim representation using ellipsis, S.T it has no limits to number of
+// dimension that the ellipsis can represent
+class EinsumToLinalgConverter : public OpConversionPattern<mhlo::EinsumOp> {
+ public:
+  using OpConversionPattern<mhlo::EinsumOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::EinsumOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter& rewriter) const final {
+    mhlo::EinsumOp::Adaptor adaptor(args);
+    auto get_rank = [](Value v) {
+      return v.getType().cast<ShapedType>().getRank();
+    };
+    auto is_scalar = [&](Value v) { return get_rank(v) == 0; };
+    auto einsum_config = op.einsum_config();
+
+    // Setting Up Commonly Used Symbols as strings
+    const std::string kArrow = "->";
+    const std::string kComma = ",";
+    const std::string kEllipsis = "...";
+
+    // With the assumption of binary Input Operand and Single Output
+    // Get the inputs and output operands' indices
+    // einsum_config = "lhs_loop,rhs_loop->out_loop"
+    std::size_t pos_arrow = einsum_config.find(kArrow);
+    std::size_t pos_comma = einsum_config.find(kComma);
+
+    StringRef lhs_loop = einsum_config.substr(0, pos_comma);
+    StringRef rhs_loop = einsum_config.substr(
+        pos_comma + kComma.size(), pos_arrow - (pos_comma + kComma.size()));
+    StringRef out_loop = einsum_config.substr(pos_arrow + kArrow.size());
+
+    // Check for Invalid Configs
+    // 1.Check that there is only maximum 2 inputs
+    // 2.Check that there is only maximum 1 output
+    // 3.Check that there is 1 kArrow
+    if (rhs_loop.find(kComma) != std::string::npos ||
+        out_loop.find(kComma) != std::string::npos ||
+        out_loop.find(kArrow) != std::string::npos) {
+      return rewriter.notifyMatchFailure(op, "Invalid Einsum Config!");
+    }
+
+    // Find result type, if on tensors.
+    ShapedType result_ty;
+    result_ty = this->typeConverter->convertType(op->getResultTypes().front())
+                    .template dyn_cast<RankedTensorType>();
```

Comment:
```suggestion
    auto result_ty = this->typeConverter->convertType(op->getResultTypes().front())
                    .dyn_cast<RankedTensorType>();
```

I think we can write it this way

```cpp
auto result_ty = this->typeConverter->convertType(GetHloOpResultType(</*isLHLO=*/false>(op)).dyn_cast<RankedTensorType>();
```

Reply:
done